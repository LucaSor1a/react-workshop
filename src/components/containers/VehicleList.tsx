import Item from "components/vehicles/Item";
import React, { useEffect, useState } from "react";
import Vehicle from "models/Vehicle";
import { Button } from "react-bootstrap";
import { get as getStock } from "services/stock";
import { Link } from "react-router-dom";
import { Spinner } from "react-bootstrap";

const VehicleList: React.FC = () => {
  const [vehicles, setVehicles] = useState<Vehicle[] | null>(null);

  async function fetchVehicles() {
    const vehicles = await getStock();

    setVehicles(vehicles);
  }

  useEffect(() => {
    fetchVehicles();
  }, []);

  if (!vehicles) {
    return (
      <Spinner animation="border" role="status">
        <span className="sr-only">Loading...</span>
      </Spinner>
    );
  }

  if (!vehicles.length) {
    return <div>Estamos sin stock. Contactanos al 0800-666-123</div>;
  }

  return (
    <ul className="list-unstyled" data-testid="vehicle-list">
      {vehicles.map((vehicle) => (
        <Item vehicle={vehicle} key={vehicle.id}>
          <Link to={`/reservar/${vehicle.id}`}>
            <Button variant="primary">Reservar online</Button>
          </Link>
        </Item>
      ))}
    </ul>
  );
};

export default VehicleList;
