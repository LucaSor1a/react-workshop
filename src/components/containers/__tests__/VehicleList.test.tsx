import axios from "axios";
import MockAdapter from "axios-mock-adapter";
import VehicleList from "../VehicleList";
import React from "react";
import { MemoryRouter as Router } from "react-router-dom";
import { render, waitForElement } from "@testing-library/react";

const mock = new MockAdapter(axios);

const response = [
  {
    id: 1,
    brand: "Ford",
    model: "Mondeo",
    year: 2020,
    price: 3000000,
    description:
      "La unión perfecta entre tecnología, sofisticación, rendimiento y eficiencia, diseño innovador e ingeniería de precisión, el Mondeo está creado para inspirar.",
  },
];

beforeEach(() => {
  mock.onGet("/api/v1/stock").reply(200, { data: response });
});

test("Renders a vehicle list", async () => {
  const { container, getByTestId, getAllByTestId } = render(
    <Router>
      <VehicleList />
    </Router>
  );

  await waitForElement(() => getByTestId("vehicle-list"));

  expect(container).toMatchSnapshot();

  expect(getAllByTestId("vehicle-item")).toHaveLength(1);
});
