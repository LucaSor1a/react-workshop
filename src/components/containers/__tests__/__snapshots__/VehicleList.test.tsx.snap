// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders a vehicle list 1`] = `
<div>
  <ul
    class="list-unstyled"
    data-testid="vehicle-list"
  >
    <li
      class="vehicle__item media"
      data-testid="vehicle-item"
    >
      <img
        alt="Mondeo"
        class="mr-3"
        data-testid="vehicle-image"
        height="auto"
        src="/img/ford/mondeo.jpg"
        width="120"
      />
      <div
        class="media-body"
      >
        <h5>
          Ford
           
          Mondeo
        </h5>
        <p>
          La unión perfecta entre tecnología, sofisticación, rendimiento y eficiencia, diseño innovador e ingeniería de precisión, el Mondeo está creado para inspirar.
        </p>
        <p>
          ARS 3,000,000.00
        </p>
        <button
          class="btn btn-primary"
          type="button"
        >
          Reservar online
        </button>
      </div>
    </li>
  </ul>
</div>
`;
